lib/ae_mdw/db/sync/generations_loader.ex

```diff
@@ -0,0 +1,126 @@
+  alias AeMdw.Log
+  alias AeMdw.Node
+
+  @max_load 100
+  @max_load_per_turn 30
+
+  @spec start_link([]) :: GenServer.on_start()
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
+  end
+
+  @impl GenServer
+  def init(:ok) do
+    {:ok, %{ranges: [], height_synced: -1}}
+  end
+
+  @spec load(Range.t()) :: :ok
+  def load(first..last) when first > last do
+    GenServer.cast(__MODULE__, {:load, last..first})
+  end
+
+  def load(first..last) do
+    GenServer.cast(__MODULE__, {:load, first..last})
+  end
+
+  @spec notify_sync(Blocks.height()) :: :ok
+  def notify_sync(height_synced) do
+    GenServer.cast(__MODULE__, {:notify_sync, height_synced})
+  end
+
+  @impl GenServer
+  def handle_cast({:load, first..last}, %{ranges: []} = state) do
+    {generation, mb_events_map} = do_load(first)
+    GenerationsCache.add(generation, mb_events_map)
+
+    if first == last do
+      {:noreply, state}
+    else
+      new_state = %{state | ranges: [(first + 1)..last]}
+      {:noreply, new_state, {:continue, :load}}
+    end
+  end
+
+  @impl GenServer
+  def handle_cast({:load, range}, %{ranges: ranges} = state) do
+    new_state = %{state | ranges: ranges ++ [range]}
+    {:noreply, new_state, {:continue, :load}}
+  end
+
+  @impl GenServer
+  def handle_cast({:notify_sync, height_synced}, state) do
+    {:noreply, %{state | height_synced: height_synced}}
+  end
+
+  @impl GenServer
+  def handle_continue(
+        :load,
+        %{ranges: [range | next_ranges], height_synced: height_synced} = state
+      ) do
+    Log.info("load #{inspect([range | next_ranges])}")
+
+    load_range = %Range{
+      first: range.first,
+      last: last_to_load(range, height_synced)
+    }
+
+    Enum.each(load_range, fn height ->
+      {generation, mb_events_map} = do_load(height)
+      GenerationsCache.add(generation, mb_events_map)
+    end)
+
+    new_state =
+      if load_range.last == range.last do
+        %{state | ranges: next_ranges}
+      else
+        next_range = %Range{first: load_range.last + 1, last: range.last}
+        %{state | ranges: [next_range | next_ranges]}
+      end
+
+    if new_state.ranges == [] do
+      {:noreply, new_state}
+    else
+      {:noreply, new_state, {:continue, :load}}
+    end
+  end
+
+  #
+  # Private functions
+  #
+  defp last_to_load(first..last, -1) do
+    Enum.min([last, first + @max_load_per_turn])
+  end
+
+  defp last_to_load(first..last, height_synced) do
+    Enum.min([
+      Enum.min([last, first + @max_load_per_turn]),
+      height_synced + @max_load
+    ])
+  end
+
+  defp do_load(height) do
+    {key_block, micro_blocks} = Node.Db.get_blocks(height)
+
+    generation = %Generation{
+      height: height,
+      key_block: key_block,
+      micro_blocks: micro_blocks
+    }
+
+    mb_events_map =
+      micro_blocks
+      |> Enum.with_index()
+      |> Enum.reduce(%{}, fn {mblock, mbi}, acc ->
+        events = AeMdw.Contract.get_grouped_events(mblock)
```

Comment:
This shall be moved to a dedicated events cache since it's a kind of dry-run on contracts which slows down the sync.